File: internal/secret/secret.go
// Package secret tracks the values that must never reach storage. Every value
// resolved from {{ env.* }} is registered here (ADR 0005: credentials come
// from the environment, not a bespoke secrets file), and the capture path
// scrubs payloads, failure details, and logs through it before anything is
// persisted.
package secret

import (
	"sort"
	"strings"
	"sync"
)

const Placeholder = "[redacted]"

// Set is a thread-safe collection of secret values. VUs share one Set per run,
// so it guards every mutation and read.
type Set struct {
	mu     sync.RWMutex
	values map[string]struct{}
}

func NewSet() *Set {
	return &Set{values: map[string]struct{}{}}
}

// Add registers a secret. Empty values are ignored so they cannot turn every
// redaction into a no-op match.
func (s *Set) Add(v string) {
	if v == "" {
		return
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.values[v] = struct{}{}
}

// Redact replaces every registered secret in text with the placeholder.
func (s *Set) Redact(text string) string {
	for _, v := range s.ordered() {
		text = strings.ReplaceAll(text, v, Placeholder)
	}
	return text
}

func (s *Set) RedactBytes(b []byte) []byte {
	return []byte(s.Redact(string(b)))
}

func (s *Set) Contains(v string) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	_, ok := s.values[v]
	return ok
}

// Values returns the registered secrets, longest-first.
func (s *Set) Values() []string { return s.ordered() }

// ordered returns the secrets longest-first, so a secret that contains another
// is replaced before its substring leaves a placeholder mid-value.
func (s *Set) ordered() []string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	vs := make([]string, 0, len(s.values))
	for v := range s.values {
		vs = append(vs, v)
	}
	sort.Slice(vs, func(i, j int) bool { return len(vs[i]) > len(vs[j]) })
	return vs
}

File: sdk-python/src/flowbench/secret.py
"""Tracks values that must never reach a stored run artifact.

Mirrors internal/secret/secret.go: every value resolved from ctx.env during
live execution is registered here, and every failure detail, span payload,
and meta field is scrubbed through it before anything is written.
"""

PLACEHOLDER = "[redacted]"


class SecretSet:
  def __init__(self):
    self._values = set()

  def add(self, value):
    if value == "":
      return
    self._values.add(value)

  def redact(self, text):
    for value in self._ordered():
      text = text.replace(value, PLACEHOLDER)
    return text

  def redact_bytes(self, data):
    return self.redact(data.decode("utf-8", errors="replace")).encode("utf-8")

  def contains(self, value):
    return value in self._values

  def _ordered(self):
    return sorted(self._values, key=len, reverse=True)
